### ISSUE TITLE
Exception when using KeyboardAttachable widget

### ISSUE THREAD

**Issue by vasylmykytiuk** [drogel/keyboard_attachable#21]
Hey @drogel,
Thank you for your great efforts on this, really appreciate it!

I'm trying to use keyboard_attachable in my app and looks like I found the bug.
```
flutter: ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
flutter: The following assertion was thrown building SizeTransition(animation: AnimationController#e3d5d(▶
flutter: 0.852)➩Interval(-2.1699839633447886⋯1.0), dirty, state: _AnimatedState#c78fd):
flutter: 'package:flutter/src/animation/curves.dart': Failed assertion: line 180 pos 12: 'begin >= 0.0': is
flutter: not true.
```

Looks like the problem is in this line: https://github.com/drogel/keyboard_attachable/blob/master/lib/src/keyboard_attachable.dart#L143

The _animationBegin is calculated to -2.1699839633447886 (that's why it fails), _bottomInset to 877.3479066666666.

This is the actual screenshot.
![telegram-cloud-photo-size-2-5456393729662367016-y](https://user-images.githubusercontent.com/3658971/126355133-d7e1c72b-3850-4056-8622-c683a0e7c388.jpg)

And this is how it should be.
![telegram-cloud-photo-size-2-5456393729662367018-y](https://user-images.githubusercontent.com/3658971/126355617-b35c858c-f9a7-4dbf-9958-0cca2abf2ee8.jpg)

Could you please take a look?

Thanks in advance.

**Comment by drogel** [comment]
Hi @vasylmykytiuk, 

Thanks for using the package and for such a detailed description of the bug! It really helps.

From version `2.0.1` and on, the `KeyboardAttachable` widget gets the absolute position of the widget that should be attached to the soft keyboard and builds the animations depeding on that. This way you can place the attached widget virtually anywhere on the screen and the animation should just fit properly. In your case, it seems to me that the `KeyboardAttachable` widget is somehow getting the wrong position for your `Next` button, and thus it's not computing the animation interpolation points properly.

An inmediate solution for you would be to downgrade to `keyboard_attachable` version `2.0.0`. That should be enough to fix the bug for your use case. 

If that doesn't solve it, please refer to the [documentation](https://github.com/drogel/keyboard_attachable/blob/master/doc/keyboard_attachable.md) or compare your code to the [example files for the package](https://github.com/drogel/keyboard_attachable/tree/master/example/lib) to see if you are using the library correctly.

In the meantime, I need to investigate further what could be causing this issue on your case. If you could provide a minimal example in which the error is caused, that would be extremely useful for me to find the problem. I will keep you updated on this if I find anything. 

Thanks!

**Comment by kornha** [comment]
Similar issue here; trying to use a text editor which launches the keyboard, as such:

```
@override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      resizeToAvoidBottomInset: false,
      navigationBar: CupertinoNavigationBar(),
      child: FooterLayout(
        footer: KeyboardAttachableFooter(),
        child: Column(
          children: [
            Expanded(
              child: Container(
                margin: EdgeInsets.symmetric(horizontal: Margins.full),
                child: QuillEditor(
                  controller: _controller,
                  focusNode: FocusNode(),
                  scrollController: ScrollController(),
                  scrollable: true,
                  padding: EdgeInsets.zero,
                  autoFocus: true,
                  readOnly: false,
                  expands: true,
                  keyboardAppearance: CupertinoTheme.brightnessOf(context),
                  customStyleBuilder: (a) =>
                      CupertinoTheme.of(context).textTheme.textStyle,
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}
```
and footer is:

```
/// Builds a footer that animates its bottom space when the keyboard is shown.
class KeyboardAttachableFooter extends StatelessWidget {
  @override
  Widget build(BuildContext context) => KeyboardAttachable(
        backgroundColor: CupertinoColors.activeBlue,
        child: Container(
          padding: const EdgeInsets.all(16),
          color: CupertinoColors.activeBlue,
          child: const CupertinoTextField(),
        ),
      );
}
```
But I get this error:
════════ Exception caught by widgets library ═══════════════════════════════════
'package:flutter/src/animation/curves.dart': Failed assertion: line 180 pos 12: 'begin >= 0.0': is not true.
The relevant error-causing widget was
KeyboardAttachable
lib/…/home/compose.dart:56
════════════════════════════════════════════════════════════════════════════════

**Comment by kornha** [comment]
Note I tried downgrading to 2.0.0 but the issue persists.

**Comment by xjodoin** [comment]
Any update on the issue?

**Comment by drogel** [comment]
Hi all, 

I have spent the whole day trying to narrow down this issue. It was hard to get the exact conditions under which this thing replicates, but I have found that it happens with certain types of layouts, when there is a `SafeArea` with `maintainBottomViewPadding` set to `false` (which is its default value). The problem is that Flutter hides the safe area bottom padding when the keyboard is shown, and it adds the safe area bottom padding again when the keyboard is dismissed, unless `maintainBottomViewPadding` is set to `true`.

I am currently testing a bugfix for this issue what would prevent this kind of errors from happening. It also polishes the dismissal animation of the keyboard attachable widget for certain kinds of layouts. However, in order for the animations to be smooth with this bugfix, `SafeArea.maintainBottomViewPadding` must be `true`. Otherwise the animations display weird jumps and snaps. I will try to make it as clear as possible in the documentation anyway.

**TL;DR**: if there are `SafeArea`s involved in the layouts in which you see this error, try setting `SafeArea.maintainBottomViewPadding` to `true`. It may fix the issue, depending on the specifics of your layout. I will be shipping a proper bugfix for this soon.

Thanks a lot!

**Comment by drogel** [comment]
Hi all, 

I have just released a new version of `keyboard_attachable` (v2.1.0) that addresses the crashes reported in this issue. 

I have also updated the documentation and examples on how to properly use `SafeArea` with `KeyboardAttachable`. Remember to set `SafeArea.maintainBorromViewPadding` to `true` and `Scaffold.resizeToAvoidBottomInsets` to `false` in your layouts, and the `KeyboardAttachable` should be able to do the rest, like so:

```dart
/// Builds a [Scaffold] that lays out a footer at the bottom of the page.
class KeyboardAttachablePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        backgroundColor: Colors.blue,
        resizeToAvoidBottomInset: false,
        appBar: AppBar(title: const Text("Keyboard Attachable demo")),
        body: SafeArea(
          maintainBottomViewPadding: true,
          child: FooterLayout(
            footer: KeyboardAttachableFooter(),
            child: ColorsList(),
          ),
        ),
      );
}
```

I am leaving this issue open for a while. If you still see this kind of problems happening, don't hesitate to comment 😄 .

Thanks!
